libiobuf - output service for job shell

chu11 opened the pull request:
This is work in progress for #2201.  I'm thinking that something that isn't very feature-ful can go into the tree as a temporary step.  This first step is very minor as the only thing is supports is allowing callers to write repeatedly to the service, the service buffers all of the data, and a caller can read all of the data back at some point later.  That's all it provides right now.  Perhaps before a merge, I could minimally support:

- tests using `flux exec`
- read from an offset / get data length
- configure buffer size

before supporting some of the more advanced things mentioned in #2201.

Anyways, I wanted to atleast post this small WIP to get comments on a few subtleties.

- I chose the name for this library as `libiobuf`, which is the best name I could come up with but it doesn't sound great.  Any ideas on something better?  Technically, it could be "databuf" as it is data agnostic.

- the library is added to `libflux_internal`.  Is it something we might want to make public so stuff outside of the project would use it?  And thus prefix all functions with `flux_`.

chu11 commented:
oh yeah

- log errors via `flux_log_error` or `log_err`?  It's a library ... but it's also setting up services with the broker.  I figure `flux_log*` for the callbacks and services, `log_err` (or no logging) for the client side stuff like `iobuf_write()`.

garlick commented:
Maybe a callback for logging?

On Wed, Jun 26, 2019, 2:57 PM Al Chu <notifications@github.com> wrote:

> oh yeah
>
>    - log errors via flux_log_error or log_err? It's a library ... but
>    it's also setting up services with the broker. I figure flux_log* for
>    the callbacks and services, log_err (or no logging) for the client
>    side stuff like iobuf_write().
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/flux-framework/flux-core/pull/2208?email_source=notifications&email_token=AABJPW4JNFPXA4KXJAB77JDP4PQ6HA5CNFSM4H3US2F2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODYU55TA#issuecomment-506060492>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABJPW5OHZRLBFMXR76SR3LP4PQ6HANCNFSM4H3US2FQ>
> .
>

garlick commented:
libiobuf seems fine to me.

One suggestion: rename `iobuf_setup()` and `iobuf_teardown()` to `iobuf_server_create()` and `iobuf_server_destroy()`.

I did notice a leftover comment about KVS dtree in one of the files.

On logging:  In general, a library shouldn't really be logging anything IMHO - it should return an appropriate level of detail to the caller and let the caller decide what to do.  Easier said than done sometimes.   On the client side, I think most errors could simply be handled with an errno.

On the server side, rather than completely hiding the message handler, would it make sense to let the user register read and write callbacks that have access to encode/decode functions for the payloads, and then do what they want with the data?  That would allow the user to detect failure at critical points and take action (like logging), and also allow for someone to store data somewhere other than memory in a different implementation...

garlick commented:
Offhand I don't see a way for a writer to indicate EOF.  Do we need that?

How does the server (e.g. in rank 0 shell) know when it's safe to tear down the I/O service?  There might be writes in flight.  (I don't have the answer - just wanted to pose the question for discussion).

Also: we should think a bit about what abnormal shutdown cases need to be handled by the shells, as opposed to the exec service, and how that affects the preservation of I/O.

chu11 commented:
> On the server side, rather than completely hiding the message handler, would it make sense to let the user register read and write callbacks that have access to encode/decode functions for the payloads, and then do what they want with the data?

Hadn't really thought about that.  My initial thought is if that API is overly complex for this particular service?

> Offhand I don't see a way for a writer to indicate EOF. Do we need that?

Yup, we will.  I had put that off for later.  

> How does the server (e.g. in rank 0 shell) know when it's safe to tear down the I/O service? There might be writes in flight. (I don't have the answer - just wanted to pose the question for discussion).

Yeah, we'll need some type of "active" and "inactive" count later on.  Thinking about it, perhaps I shouldn't create a buffer on the first "write" that occurs to a "stream-rank" combo.  There should be a "register buffer", "finished/EOF buffer", and "delete buffer" type of service callbacks.  Which based on these, leads to the active vs inactive count.

chu11 commented:
Just did a push with a more complete first run of the library service.  Still more todo, but possible merge candidate with a few more tiny things added.  So the service right now has callbacks to:

- create a buffer (optional, first write will also create)
- write to a buffer
- read from the buffer (all contents, or via offset & count)
- get the length of contents the buffer
- delete a buffer
- mark eof on a buffer
- get count of all buffers
- get count of all buffers marked with EOF

smaller TODOs

- `flux-exec` tests
- option for error logging or just remove logging

bigger TODOs / to think about

- dump buffer to KVS / file / something on ... exit / callback / something
- dump buffer when buffer crosses threshold
- option - do not allow more buffers to be created
- option - do not allow `write` to create buffer if one doesn't exist (idea: caller wants to create all "known" buffers at beginning, does not want callers of `write` to accidentally create a buffer)

grondo commented:
> Yeah, we'll need some type of "active" and "inactive" count later on. 

In the interest of expediency, perhaps this service should be tailored to the job shell use case for now, and then expand to other cases as they come up.

A user like the job shell will want to create the service with a known number of streams or buffers, and then be notified or have the service automatically terminate only when all known streams get EOF. (Though of course the actual buffer creation could be delayed until first write)

grondo commented:
Actually, how did you envision the job shell would use this service (maybe with a psuedocode example)? That might help with the review.

chu11 commented:
> Actually, how did you envision the job shell would use this service (maybe with a psuedocode example)? That might help with the review.

I hadn't thought about it too much right, since I was just trying to get the core of the service built up.

Thinking about it, I imagine the stdout/stderr callbacks from libsubprocess (`on_stdout` and `on_stderr`) would forward onto the `libiobuf` service by calling `iobuf_write()`.  And when they get an `EOF`, to call `iobuf_eof()`.  Or perhaps the EOF could be called via the `on_completion` callback (that might be better).

The "end" scenario I wasn't too sure about yet.  When the job-shell realizes all sub-processes are complete, should it call a "dump the IO to the KVS" kind of function?  Or should there be a flag that automatically tells the iobuf service to do this whenever EOF is reached?

grondo commented:
> Or perhaps the EOF could be called via the on_completion callback (that might be better).

Since a task may close an output stream well before it exits, I'd send EOF as soon as it is available.

>  When the job-shell realizes all sub-processes are complete, should it call a "dump the IO to the KVS" kind of function? Or should there be a flag that automatically tells the iobuf service to do this whenever EOF is reached?

For now it might be easiest to just have a callback that is triggered when all streams have EOF. The job shell can register this callback to dump the output to the kvs. However, in order to support this the IO service has to support creation with a predetermined number of streams. This should probably be the default behavior, as I don't see how output could ever be "finished" if you don't know the number of EOFs for which you are waiting.

grondo commented:
> Edit/Adding: For this round of libiobuf, I was thinking of just trying to add the "basics" of the service, i.e. buffer, write, read, EOF, etc. None of the bigger things listed above. Perhaps how those advanced parts will come about as #2211 is done.

Yeah, I was just implying that the "basics" probably should be the bare minimum of what is needed for the first round of the job shell: (create an IO service for N stdout/err streams, trigger callback when all EOFs received)

chu11 commented:
> However, in order to support this the IO service has to support creation with a predetermined number of streams. This should probably be the default behavior, as I don't see how output could ever be "finished" if you don't know the number of EOFs for which you are waiting.

Ahh, good point.  Do you think the service should be initialized with such a number of buffers?  If the user attempts to create more buffers than that, it's an error on their part.

garlick commented:
Would it help to rebase on job shell pr and implement stdio to kvs?  (Maybe
job shell is too subject to change at this point?)

On Tue, Jul 9, 2019, 2:12 PM Al Chu <notifications@github.com> wrote:

> However, in order to support this the IO service has to support creation
> with a predetermined number of streams. This should probably be the default
> behavior, as I don't see how output could ever be "finished" if you don't
> know the number of EOFs for which you are waiting.
>
> Ahh, good point. Do you think the service should be initialized with such
> a number of buffers? If the user attempts to create more buffers than that,
> it's an error on their part.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/flux-framework/flux-core/pull/2208?email_source=notifications&email_token=AABJPW3R5ROLPMLRUARE3YDP6T5MHA5CNFSM4H3US2F2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODZRRVLY#issuecomment-509811375>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABJPW2URINSE3UWGXID7YDP6T5MHANCNFSM4H3US2FQ>
> .
>

grondo commented:
> Do you think the service should be initialized with such a number of buffers? If the user attempts to create more buffers than that, it's an error on their part.

I suppose yes. To get this functionality into the job shell as quickly as possible, I would say the simplest library possible that allows the following:

 * "rank 0" job shell starts the "io service" for N tasks, with 1 buffer for stdout/stderr per task either implied or explicitly set.
 * each job shell then sends task output to the service using provided api until eof.
 * When all EOFs have been received by the "io service" a callback in job shell is triggered (which can then dump all buffers to kvs for now)

If a task output message comes into the io service for a stream that shouldn't exist, then that should probably throw an error.

After this is working, we would have some experience and groundwork to figure out how tools like `flux run` or `flux job attach` could then "attach" to one or more I/O streams so we can get "live" output (and eventually send input) to get closer to parity with the old system.

This is just how I pictured this initial output service working. I wanted to make sure I  wasn't missing something you and perhaps @garlick already had in mind, though, since I haven't been around.

chu11 commented:
> Would it help to rebase on job shell pr and implement stdio to kvs? (Maybe job shell is too subject to change at this point?)

Nah, I think this library is a still in a bit of TBD mode as well.

chu11 commented:
> If a task output message comes into the io service for a stream that shouldn't exist, then that should probably throw an error.

Do you think the job-shell should be responsible for pre-creating the names of the buffers?  Likely something with jobid & stream & rank?  By default, I had `iobuf_write ()` create a buffer if it didn't exist.

grondo commented:
> Do you think the job-shell should be responsible for pre-creating the names of the buffers? Likely something with jobid & stream & rank? By default, I had iobuf_write () create a buffer if it didn't exist.

Yeah, that aspect of the design is where I wasn't sure how the library was meant to be used by the job shell. I probably just need to look at the implementation further, but got distracted looking at the job shell PR.

If buffers are autocreated by iobuf_write () then you run into a problem where an instance of the iobuf server doesn't know beforehand how many streams for which it needs to wait on EOF. I guess it could work if the job shell checks the number of EOFs on each EOF, so maybe we should just try it out in the interest of moving forward?

chu11 commented:
> If buffers are autocreated by iobuf_write () then you run into a problem where an instance of the iobuf server doesn't know beforehand how many streams for which it needs to wait on EOF. I guess it could work if the job shell checks the number of EOFs on each EOF, so maybe we should just try it out in the interest of moving forward?

One idea I thought of is simply configuring some type of "max" buffers count.  Trigger the callback after receiving that number of EOFs (or perhaps it'd simply be an EOF count).

grondo commented:
> One idea I thought of is simply configuring some type of "max" buffers count. Trigger the callback after receiving that number of EOFs (or perhaps it'd simply be an EOF count).

Yeah, that seems like a simple solution for now.

garlick commented:
To summarize face to face discussion re: I/O in the job shell:
* service registration should return a future, not run synchronously
* (actually just noticed) service name should be set by the user (e.g. stdio-_jobid_)  not hardwired
* need some way to allow task I/O to accumulate in task's buffer while waiting for barrier indicating iobuf service is available (a proposed change to libsubprocess, not here)
* instead of implementing per-task character buffers on the service end, would be better to enqueue messages in a time-ordered list, then provide user with access to the list
* (extra credit) provide decode function for messages from list

chu11 commented:
just did push

- added a few extra settings, `bufsize`, `max_buffer_count`, and  `flags` (for logging) when you create the server object

- added a 'register' and 'unregister' service functions, so that registering/unregistering the service is not synchronous.  I wasn't  sure how to fit asynchronous in with the `iobuf_server_create/destroy()` calls, so I elected to splice it out into new functions.

- added an EOF callback that is called after some number of EOFs are set on streams/buffers.

- Setting an EOF previously assumed a buffer was actually created before hand.  This probably won't work with the job-shell, b/c some processes could have zero I/O but still need to send EOF at the end.  So you can set an EOF even if a buffer was never created.

- I removed the `delete` operation, as it started complicating things once I had the EOF cb.

- I have a few operations that may not be necessary, but I left them in for the time being, notably: `length`, `buffer_count`, and `buffer_eof_count`.

- If too much data goes to a buffer, I just overflow and return an error to the user.  Don't have a way to "drain" a buffer at the moment.

TODO

- instead of a cbuf, just buffer messages containing stdio in a big zlist
  - could have helper functions to read from zlist, i.e. get stdout from rank 2 
- some operations (length, buffer count, eof count, etc.) do not require an RPC, since it will only be called by the creator of the service.  Need to redo that part.
- flux-exec tests

garlick commented:
I would suggest just dropping the register/unregister and leave that to the caller.   The caller can pass the name in to `iobuf_server_create()` as it needs to be different for each job.

chu11 commented:
> (actually just noticed) service name should be set by the user (e.g. stdio-jobid) not hardwired

Doh!  I passed the wrong thing to the service register/unregister.

chu11 commented:
> I would suggest just dropping the register/unregister and leave that to the caller. The caller can pass the name in to iobuf_server_create() as it needs to be different for each job.

Ok.  So the `name` passed in by the user will the exact name of the service.  I currently manipulate the name (prefix/suffix), so have to tweak that.

chu11 commented:
do we need a way for the caller/user to know what io is available, i.e. "stdout.1", "stdout.2" are available, but "stdout.3" in not available?

garlick commented:
> do we need a way for the caller/user to know what io is available, i.e. "stdout.1", "stdout.2" are available, but "stdout.3" in not available?

I don't think for this use case.  In fact we probably don't even need a way to find messages by stream name.  We'll just iterate through all of them (or some "recent" subset for attach, potentially).

garlick commented:
One other thought: leader shell will need a message encode function and ability to directly enqueue a message (since it is also managing tasks that produce output).

chu11 commented:
So basically a non-RPC version of `iobuf_write()` to enqueue messages?

garlick commented:
Yep.

grondo commented:
> One other thought: leader shell will need a message encode function and ability to directly enqueue a message (since it is also managing tasks that produce output).

Hm, though slightly less efficient, there might be some benefit to just having the leader shell send output to its own service. For example, simpler, it would be easier to move the service somewhere else in the future, no need to special case code in the leader shell, and if we wanted to add message based filtering/monitoring it might be easier to accomplish if all messages are coming from the same place.

Not a big deal though, just mentioning it.

garlick commented:
Either way would be OK IMHO.  I can't remember if there's a limitation on sending a request to _yourself_, so maybe a quick test would be in order if you go that way.

grondo commented:
> I can't remember if there's a limitation on sending a request to yourself, so maybe a quick test would be in order if you go that way.

IIRC, the python unit test for `service.add` sends a message to itself...

chu11 commented:
> Either way would be OK IMHO. I can't remember if there's a limitation on sending a request to yourself, so maybe a quick test would be in order if you go that way.

As I thought about it, I think have local versions would be useful for testing, so I was thinking of adding it anyways.

chu11 commented:
New push re-implements the `iobuf` library to internally store a list of each write that occurs in the iobuf service.  In addition, there are both "local" and "rpc" functions for doing create, write, eof, and read.  And there is a simple iterator for iterating over all data in the list.

Some notes/thoughts

- I currently base64-to-binary when storing all data received via rpc.  This may not be very efficient unless we think the common case is to eventually store all this data at the end of the service running?

- To read the data for a specific stream/rank the entire list of data has to be iterated over.  This can be inefficient if getting stdout/stderr for a specific stream/rank is a common pattern.  Plan is to add some pointers to create a list inside a list for each stream/rank combo.  (As an aside, I don't know if there is a name for such a data structure?  A list with a bunch of lists inside it?  It's not really a skip-list.)

- The return codes for `iobuf_write()` and `iobuf_read()` are simply 0 or -1.  Could return length of write/read?  Reason I did just 0 and -1 is to be consistent between the local & rpc versions of the functions.  But could go either way.

- When someone tries to write to a stream that has already been EOFed, I return errno = `EROFS`.  Which seems wrong.  I believe with `write(2)` you would get `EPIPE` for the other end being closed or `EBADF` if the `fd` was already closed.  Both of those didn't seem right either.

- Similarly when there are too many buffers created I chose `ENFILE`.  Maybe could do better?

garlick commented:
I'll put in some time reviewing this morning when I get to work.

I do have the shell use case (at least the initial one that just collects I/O and dumps it into the KVS at the end of the job) prototyped in #2246 (see src/shell/io.c).  It's pretty simple if you want to have a look and make sure the proposed interfaces here fit more or less over there.

chu11 commented:
I think that this PR is in a form to be considered for merging now.  The most recent push added some more tests, cleaned up comments, and added some performance improvements when a reader wants a specific buffer for a stream/rank.

Only question I still have is if anyone can think of better errnos for the two issues I described above:

> - When someone tries to write to a stream that has already been EOFed, I return errno = EROFS. Which seems wrong. I believe with write(2) you would get EPIPE for the other end being closed or EBADF if the fd was already closed. Both of those didn't seem right either.
> - Similarly when there are too many buffers created I chose ENFILE. Maybe could do better?

garlick commented:
Apologies for not getting to this yesterday.  At this point I think I'd vote that #2246 goes in before this and we make sure that this library both fits that use case well and adds capability so that we can support the advanced I/O modes in #2241 when we cut over.  (It may be close now, but I haven't had much of a chance to look in detail).

One quick suggestion is to please split the large commit into library and tests at minimum.

chu11 commented:
after discussion, it was determined that this lib isn't really necessary, since it's only going to be used by the shell.  It should really just be added into the shell directly.

Parts were salvaged into libioencode (#2293)
